### ISSUE TITLE
gocql prepared all queries

### ISSUE THREAD

**Issue by GQHero** [gocql/gocql#845]
Hi,

It seems that gocql will prepare all the queries like CRUD.
But we have lots of random read and write, which may cause the cache soon run out.
Is there some ways to avoid using prepared query?

Thanks.

**Comment by vrischmann** [comment]
[This](https://sourcegraph.com/github.com/gocql/gocql@db6f35eb602dd0770a7162e0e23e5cfb1cc05b91/-/blob/session.go#L826:17) is always called when executing a query, and looks like the behavior is hard coded.

Right now it looks like your best bet would be to modify gocql directly and change `shouldPrepare` to your liking.

**Comment by Zariel** [comment]
The issue arises from the fact that the driver wont know how to serialise your types into the columns Cassandra expects without preparing, all prepare really does is cache the type information which is used when serialising the request.

You shouldn't really be thrashing the cache, if you're using queries without bind params then thats a simple fix.

**Comment by rockxsj** [comment]
the java driver support raw query, we need too

**Comment by martin-sucha** [comment]
@rockxsj what is your use case?

**Comment by rockxsj** [comment]
> @rockxsj what is your use case?

I have a data gateway, which for accept data sink request and persistent to Cassandra.  Most sink request can use prepared statement bind with params, but some complex data type I want direct execute the statement send by request.

**Comment by martin-sucha** [comment]
I've re-read this issue and it is still not clear to me when it is an issue using prepared statements. @rockxsj could you please provide a more concrete example of "some complex data type" that requires directly executing the query without preparing? Or what exactly is the issue with using the prepared statements?

**Comment by OleksiienkoMykyta** [comment]
I don't see the issue using prepared queries, it's a little overhead if it's a single-use query, but in other cases, it provides more efficiency, security, and performance. Not sure that there is something to fix.

The issue reporter hasn't responded since Dec 1, 2016.
Candidate to be closed.

**Comment by utking** [comment]
@OleksiienkoMykyta,

We have a Cassandra v4 cluster with 3 nodes carrying ~3TB of data, and switching from Java version with raw queries to Go version where everything is prepared has dramatically increased using (and flushing) the prep statements cache. The Go version significantly increased CPU usage (~90% across all the cores) read/write and resulted in OS LoadAverage from 10 to 30, while with the Java version it was 2-5 for the same workloads (switching back to the Java version, confirmed that the LoadAverage goes down immediately).
We optimized the queries to make them all prepare-able - it helped with the prepares statements' cache, but the CPU usage and read/write volume/latency are still higher and the LoadAverage is high with them (from 10 to 30).
Though it not an issue under "normal" load, a more heavier use, clearly, hurts the cluster performance considerably.
Do you need some more data about our case?

**Comment by joao-r-reis** [comment]
So after you made the queries preparable did you double check to see if you're still seeing cache evictions or is the size of the cache stable? If you're misuing prepared statements then you will run into the issue of the cache constantly hitting the limit and having to evict statements.

If you're sure you're not seeing cache evictions now and you're still not happy with the performance of the cache then the only thing that would maybe help is using a cache that isn't limited by the performance of `sync.Mutex` but we'd really need to see some data to see that the mutex is the bottleneck. And even if that was the case then we would probably just make it an interface and allow the user to provide a custom cache implementation because I don't think this is a problem that enough users will face that would justify adding an external dependency for a LRU cache library like `maypok86/otter` (maybe I could see it on a separate package).

It could also just be that the reason gocql is slower than java for your workload is unrelated to the prepared cache which is another reason why we'd like to see some data suggesting that changing the cache or even allowing statements to be unprepared would solve anything.

Personally I believe that in general prepared statements should be the way to go if you're worried about performance even on the java driver because the column metadata is not sent at all when the queries are prepared so it's less data going through the wire but it is possible that `sync.Mutex` is slow enough that it breaks this assumption.

**Comment by joao-r-reis** [comment]
Also can you check what protocol version and compression you're using? And whether they are the same for both gocql and the java driver? Gocql only supports protocol v5 with the latest release (2.0.0) while the java driver has had it for a while.

**Comment by utking** [comment]
@joao-r-reis 
Yes, we did check and we don't see cache flushes often anymore (they do occur, but very rarely). The cache size (from the Cassandra DEBUG logs) seems stable. This, in my opinion, shows that we are not misusing the cache as we did issuing raw statements in the start of the switch.

It's not that we are not happy with the performance of the cache. It's that with the [seemingly] stable cache we still see high CPU usage and read/write volumes and latency. I am not sure it is attributable to the prepared statements cache.

I am also not sure why `sync.Mutex` would be the issue since we do not observe any load (or otherwise) issues with the proxy using the driver; we observe them only on the Cassandra cluster nodes that the proxy talks to.

Go driver being slower than the Java one - we compared the two driver versions on the very same [production] workload. The processing time for workers talking to Cassandra through the proxies did not change. If the Go driver is slow, I don't really see how/what to check to confirm that.

I want to agree that prepared statement are a better way to go, but if it turns out they are causing the issue, I would prefer them conditional instead.

We tried protocols v3 (originally, because it was used with the Java version) and v4 (to check if anything will change; nothing did). I did prepare the v2.0.0-based version to try protocol v5 when we have a chance (scheduled in the end of December 2025), but I saw nothing in the changelog that would suggest it should improve the situation. Also, as it was suggested earlier, I probably will try a custom fork with the `shouldPrepare` being disabled - just to finally test the assumption that it is the unconditional prepared statements that cause the LoadAverage spikes under heavy workloads.

Thanks for your prompt response!

**Comment by joao-r-reis** [comment]
> Also, as it was suggested earlier, I probably will try a custom fork with the shouldPrepare being disabled - just to finally test the assumption that it is the unconditional prepared statements that cause the LoadAverage spikes under heavy workloads.

This will be very helpful, when you get results back from this experiment please let us know

**Comment by utking** [comment]
@joao-r-reis 
Sorry, it takes longer than expected. For today, we tested driver v2 - no improvements compared to v1. Once we have the results from forking the driver to try prepared-less statements, I'll let you know.
